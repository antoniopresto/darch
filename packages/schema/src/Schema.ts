import { RuntimeError } from '@darch/utils/lib/RuntimeError';
import { StrictMap } from '@darch/utils/lib/StrictMap';
import { expectedType } from '@darch/utils/lib/expectedType';
import { getTypeName } from '@darch/utils/lib/getTypeName';
import { invariantType } from '@darch/utils/lib/invariant';
import { simpleObjectClone } from '@darch/utils/lib/simpleObjectClone';
import { ForceString } from '@darch/utils/lib/typeUtils';

import { SchemaDefinitionInput } from './TSchemaConfig';

import {
  parseValidationError,
  ValidationCustomMessage,
} from './applyValidator';

import { validateSchemaFields } from './getSchemaErrors';
import { parseSchemaDefinition } from './parseSchemaDefinition';
import {
  FinalSchemaDefinition,
  ParseFields,
  SchemaFieldInput,
  ToFinalField,
} from './fields/_parseFields';
import { Infer } from './Infer';
import type { ObjectTypeComposer } from 'graphql-compose';
import { isBrowser } from '@darch/utils/lib/isBrowser';
import type { GraphQLInputObjectType, GraphQLObjectType } from 'graphql';
import { dynamicRequire } from '@darch/utils/lib/dynamicRequire';
import {
  isMetaFieldKey,
  MetaFieldDef,
  schemaMetaFieldKey,
} from './fields/MetaFieldField';
import type { CreateResolverOptions, Resolver } from './createResolver';

export { RuntimeError } from '@darch/utils/lib/RuntimeError';
export * from './parseSchemaDefinition';
export * from './schemaInferenceUtils';

export class Schema<DefinitionInput extends SchemaDefinitionInput> {
  private readonly __definition: any;

  get definition(): ParseFields<DefinitionInput> {
    return this.__definition;
  }

  get description() {
    return this.meta.description;
  }

  constructor(schemaDef: DefinitionInput) {
    const parsed = parseSchemaDefinition(schemaDef);
    this.__definition = parsed.definition;
  }

  get meta(): MetaFieldDef {
    return this.__definition[schemaMetaFieldKey].def;
  }

  private __setMetaData(k: keyof MetaFieldDef, value: string | number) {
    this.__definition[schemaMetaFieldKey].def[k] = value;
  }

  parse(
    input: any,
    options?: { partial?: boolean; customMessage?: ValidationCustomMessage }
  ): Infer<DefinitionInput> {
    const { customMessage } = options || {};
    const { errors, parsed } = this.safeParse(input, options?.partial);

    if (errors.length) {
      const err: any = parseValidationError(
        input,
        customMessage,
        errors.join(' \n')
      );
      err.isSchemaValidationError = true;
      err.fieldErrors = errors;
      throw err;
    }

    return parsed as any;
  }

  validate(input: any): input is Infer<DefinitionInput> {
    try {
      this.parse(input);
      return true;
    } catch (e) {
      return false;
    }
  }

  safeParse(
    input: any,
    partial?: boolean
  ): { errors: string[]; parsed: unknown } {
    const SchemaConstructor: any = this.constructor;

    const errors: string[] = [];
    const parsed: any = {};

    if (!input || typeof input !== 'object' || Array.isArray(input)) {
      throw new RuntimeError(
        `Invalid input. Expected object, found ${getTypeName(input)}`,
        {
          input,
        }
      );
    }

    Object.keys(this.definition).forEach((currField) => {
      if (isMetaFieldKey(currField)) return;

      const value = input[currField];

      const fieldDef = this.definition[currField];

      if (value === undefined && partial) {
        return;
      }

      const result = validateSchemaFields({
        createSchema: (def) => new SchemaConstructor(def),
        fieldName: currField,
        definition: fieldDef,
        value,
      });

      if (Object.prototype.hasOwnProperty.call(input, currField)) {
        parsed[currField] = result.parsed;
      }

      errors.push(...result.errors);
    });

    return { parsed, errors };
  }

  describe(
    ...descriptions:
      | [comment: string]
      | [{ [K in keyof DefinitionInput]?: string }]
  ): Schema<DefinitionInput> {
    if (descriptions.length === 1 && typeof descriptions[0] === 'string') {
      this.__setMetaData('description', descriptions[0]);
      return this;
    }

    const commentsConfig = descriptions[0];

    invariantType({ commentsConfig }, 'object', { commentsConfig });

    const definition: FinalSchemaDefinition = this.definition as any;

    Object.entries(commentsConfig).forEach(([name, comment]) => {
      invariantType(
        { [name]: definition[name] },
        'object',
        `"${name}" is not in schema definition.`
      );
      definition[name].description = comment || '';
    });

    return this;
  }

  removeField<K extends ForceString<keyof DefinitionInput>>(
    field: K | K[]
  ): OmitDefinitionFields<DefinitionInput, K> {
    const fields: string[] = Array.isArray(field) ? field : [field];
    const clone = this.clone();

    for (const k in clone.__definition) {
      if (fields.includes(k)) {
        delete clone.__definition[k];
      }
    }

    return clone as any;
  }

  addFields<T extends SchemaDefinitionInput>(
    definition: T
  ): SchemaExtendDefinition<ParseFields<DefinitionInput>, T> {
    return this.clone(definition) as any;
  }

  makeOptional<K extends ForceString<keyof DefinitionInput>>(
    fields: K | K[]
  ): Schema<MakeOptional<DefinitionInput, K>> {
    const fieldList = Array.isArray(fields) ? fields : [fields];
    const clone = this.clone();
    fieldList.forEach((key) => (clone.__definition[key].optional = true));
    return clone as any;
  }

  makeRequired<K extends ForceString<keyof DefinitionInput>>(
    fields: K | K[]
  ): Schema<MakeRequired<DefinitionInput, K>> {
    const fieldList = Array.isArray(fields) ? fields : [fields];
    const clone = this.clone();
    fieldList.forEach((key) => (clone.__definition[key].optional = false));
    return clone as any;
  }

  get __isDarchSchema(): true {
    return true;
  }

  clone(name?: string): this;

  clone<T extends SchemaDefinitionInput>(
    extend: T,
    name?: string
  ): SchemaExtendDefinition<ParseFields<DefinitionInput>, T>;

  clone<T extends Record<string, SchemaFieldInput | null>>(
    extend: (current: ParseFields<DefinitionInput>) => T,
    name?: string
  ): Schema<{ [K in keyof ExcludeNull<T>]: ExcludeNull<T>[K] }>;

  clone<K extends keyof DefinitionInput>(
    fields: K[],
    name?: string
  ): Schema<CloneFields<ParseFields<DefinitionInput>, K>>;

  clone<
    K extends keyof DefinitionInput,
    T extends Record<string, SchemaFieldInput | null>
  >(
    fields: K[],
    extend: (current: CloneFields<ParseFields<DefinitionInput>, K>) => T,
    name?: string
  ): Schema<{ [K in keyof ExcludeNull<T>]: ExcludeNull<T>[K] }>;

  clone(...args: any[]) {
    const lastArg = args[args.length - 1];

    let newId;
    if (typeof lastArg === 'string') {
      newId = args.pop();
    }

    const definitionClone = simpleObjectClone(this.definition);
    delete definitionClone[schemaMetaFieldKey];

    let extend;
    let fields;

    if (args.length === 1) {
      if (Array.isArray(args[0])) {
        fields = args[0];
      } else {
        extend = args[0];
      }
    }

    if (args.length === 2) {
      fields = args[0];
      extend = args[1];
    }

    if (Array.isArray(fields)) {
      for (const k in definitionClone) {
        if (!fields.includes(k)) {
          delete definitionClone[k];
        }
      }
    }

    const def =
      typeof extend === 'function'
        ? extend(definitionClone)
        : typeof extend === 'object'
        ? { ...definitionClone, ...extend }
        : definitionClone;

    Object.keys(def).forEach((k) => {
      if (!def[k]) {
        delete def[k];
      }
    });

    const schema = createSchema(def);

    if (newId) {
      schema.identify(newId);
    }

    return schema as any;
  }

  get id() {
    return this.meta.id;
  }

  identify<ID extends string>(id: ID): this & { id: ID } {
    if (id && id === this.id) return this as any;

    if (this.id) {
      throw new Error(
        `Trying to replace existing id "${this.id}" with "${id}". You can clone it to create a new Schema.`
      );
    }

    expectedType({ id }, 'string', 'truthy');

    if (Schema.register.has(id) && Schema.register.get(id) !== this) {
      console.error(`Schema with id "${id}" already registered.`);
    }

    this.__setMetaData('id', id);
    Schema.register.set(id, this);

    return this as any;
  }

  static register = new StrictMap();

  private _otc: ObjectTypeComposer | undefined;

  entity = (name?: string): ObjectTypeComposer => {
    if (this._otc) return this._otc;
    if (isBrowser() || typeof module?.require !== 'function') {
      throw new Error('GraphQL transformation is not available in browser.');
    }

    if (name) {
      this.identify(name);
    }

    if (!this.id) {
      throw new RuntimeError(
        'Should schema.identify() before converting to Graphql.' +
          '\nYou can call schema.clone() to choose a different identification.',
        { 'used definition': this.definition }
      );
    }

    const { schemaToGQL } = dynamicRequire('./schemaToGQL', module);

    return (this._otc = schemaToGQL(this.id, this.definition));
  };

  private _graphqlType: GraphQLObjectType | undefined;
  graphqlType = (): any => {
    if (this._graphqlType) return this._graphqlType;
    return (this._graphqlType = this.entity().getType());
  };

  private _graphqlInputType: GraphQLInputObjectType | undefined;
  graphqlInputType = (): any => {
    if (this._graphqlInputType) return this._graphqlInputType;
    return (this._graphqlInputType = this.entity().getInputType());
  };

  createResolver = <
    Args extends SchemaDefinitionInput,
    Context = unknown,
    Source = unknown
  >(
    options: Omit<CreateResolverOptions<this, Args, Context, Source>, 'type'>
  ): Resolver<this, Args, Context, Source> => {
    return Schema.serverUtils().createResolver({
      ...options,
      type: this,
    });
  };

  static serverUtils() {
    if (isBrowser() || typeof module?.require !== 'function') {
      throw new Error('Server utils are not available on browser.');
    }

    return {
      graphqlCompose: dynamicRequire(
        'graphql-compose',
        module
      ) as typeof import('graphql-compose'),

      graphql: dynamicRequire('graphql', module) as typeof import('graphql'),

      createResolver: dynamicRequire('./createResolver', module)
        .createResolver as typeof import('./createResolver')['createResolver'],
    };
  }

  static createResolver = <
    TypeDef extends SchemaFieldInput | Schema<SchemaDefinitionInput>,
    Args extends SchemaDefinitionInput,
    Context = unknown,
    Source = unknown
  >(
    options: CreateResolverOptions<TypeDef, Args, Context, Source>
  ): Resolver<TypeDef, Args, Context, Source> => {
    return Schema.serverUtils().createResolver(options);
  };

  static async reset() {
    Schema.serverUtils().graphqlCompose.schemaComposer.clear();
    Schema.register.clear();
  }
}

export const DarchSchema = Schema;

export function createSchema<
  DefinitionInput extends Readonly<SchemaDefinitionInput>
>(fields: Readonly<DefinitionInput>): Schema<DefinitionInput>;

export function createSchema<
  DefinitionInput extends Readonly<SchemaDefinitionInput>
>(name: string, fields: DefinitionInput): Schema<DefinitionInput>;

export function createSchema<
  DefinitionInput extends Readonly<SchemaDefinitionInput>
>(
  ...args: [string, DefinitionInput] | [DefinitionInput]
): Schema<DefinitionInput> {
  const fields = args.length === 2 ? args[1] : args[0];
  const id = args.length === 2 ? args[0] : undefined;

  const schema = new Schema<DefinitionInput>(fields);
  if (id) return schema.identify(id);

  return schema;
}

export { createSchema as createType };

type OmitDefinitionFields<T, Keys extends string> = T extends {
  [K: string]: any;
}
  ? Schema<{ [K in keyof T as K extends Keys ? never : K]: T[K] }>
  : never;

type SchemaExtendDefinition<T, Ext> = T extends { [K: string]: any }
  ? Ext extends { [K: string]: any }
    ? Schema<{
        [K in keyof (T & Ext)]: (T & Ext)[K];
      }>
    : never
  : never;

type CloneFields<T, Keys> = T extends {
  [K: string]: any;
}
  ? { [K in keyof T as K extends Keys ? K : never]: T[K] }
  : never;

type MakeOptional<T, Keys extends string> = T extends {
  [K: string]: any;
}
  ? {
      [K in keyof T]: K extends Keys
        ? ToFinalField<T[K]> extends infer Obj
          ? { [K in keyof Obj]: K extends 'optional' ? true : Obj[K] }
          : never
        : T[K];
    }
  : never;

type MakeRequired<T, Keys extends string> = T extends {
  [K: string]: any;
}
  ? {
      [K in keyof T]: K extends Keys
        ? ToFinalField<T[K]> extends infer Obj
          ? { [K in keyof Obj]: K extends 'optional' ? false : Obj[K] }
          : never
        : T[K];
    }
  : never;

type ExcludeNull<T> = {
  [K in keyof T as [T[K]] extends [null]
    ? never
    : [T[K]] extends [undefined]
    ? never
    : K]: Exclude<T[K], null>;
};
