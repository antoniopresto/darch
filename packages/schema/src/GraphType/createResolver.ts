import {
  GraphQLField,
  GraphQLFieldConfig,
  GraphQLObjectType,
  GraphQLResolveInfo,
} from 'graphql';

import { Infer } from '../Infer';
import { FieldDefinitionConfig } from '../TObjectConfig';
import {
  InferField,
  ObjectDefinitionInput,
  ToFinalField,
} from '../fields/_parseFields';

import { createType, GraphType } from './GraphType';
import { getInnerType } from './getQueryExamples';

export function createResolver<
  Type extends FieldDefinitionConfig,
  Args extends ObjectDefinitionInput
>(
  options: ResolverConfig<any, any, Type, Args>
): Resolver<any, any, Type, Args> {
  const { args, name, kind = 'query', resolve, type, ...rest } = options;

  if (GraphType.resolvers.has(name)) {
    return GraphType.resolvers.get(name);
  }

  const payloadType = (GraphType.is(type)
    ? type
    : createType(`${name}Payload`, type)) as unknown as GraphType<any>;

  const gqlType = payloadType.graphQLType();

  const argsType = createType(
    `${name}Input`,
    isPossibleArgsDef(args) ? { object: args } : 'record'
  ) as unknown as GraphType<any>;

  const argsGQLType = argsType.graphQLInputType({ name: `${name}Input` });
  const innerArgsGQLType = getInnerType(argsGQLType);

  const resolveFunction: any = async function typeCheckResolveWrapper(
    source,
    args: any,
    context: any,
    info: any
  ): Promise<any> {
    args = argsType
      ? argsType.parse(args, {
          customMessage: (_, error) => {
            return `Invalid input provided to resolver "${name}":\n ${error.message}`;
          },
        } as any)
      : args;

    const result = await resolve(source, args, context, info);

    return payloadType.parse(
      result,
      (_, error) => `Invalid output from resolver "${name}": ${error.message}`
    );
  };

  const result: AnyResolver = {
    ...rest,
    kind,
    name,
    resolve: resolveFunction,
    args:
      'getFields' in innerArgsGQLType.innerType
        ? innerArgsGQLType.innerType.getFields()
        : {},
    type: gqlType,

    // Resolver fields - not in GraphQLFieldConfig
    argsDef: args,
    typeDef: payloadType.definition,
    payloadType,
    argsType,
    asObjectField: (_name = name): GraphQLField<any, any> => {
      const temp = new GraphQLObjectType({
        name: 'temp',
        fields: {
          [_name]: result,
        },
      });

      return temp.getFields()[_name];
    },
    __isResolver: true,
    __isRelation: false,
    __relatedToGraphTypeId: '',
    __graphTypeId: payloadType.id,
  };

  GraphType.resolvers.set(name, result);

  return result as any;
}

type InferArgs<ArgsDef> = InferField<
  ToFinalField<{
    object: ArgsDef;
  }>
>;

export interface ResolverConfig<Context, Source, TypeDef, ArgsDef>
  extends Omit<GraphQLFieldConfig<any, any>, 'resolve' | 'args' | 'type'> {
  name: string;
  kind?: 'query' | 'mutation' | 'subscription';
  args?: ArgsDef;
  type: TypeDef;

  resolve(
    root: Source,
    args: InferArgs<ArgsDef>,
    context: Context,
    info: GraphQLResolveInfo
  ): Infer<ToFinalField<TypeDef>> | Promise<Infer<ToFinalField<TypeDef>>>;
}

export type AnyResolver = Resolver<any, any, any, {}>;

export interface Resolver<Context, Source, TypeDef, ArgsDef>
  extends Omit<GraphQLFieldConfig<any, any>, 'resolve' | 'args' | 'type'> {
  __isResolver: true;
  __isRelation: boolean;
  __graphTypeId: string;
  __relatedToGraphTypeId: string;

  resolve(
    root: Source,
    args: InferArgs<ArgsDef>,
    context: Context,
    info: GraphQLResolveInfo
  ): Infer<ToFinalField<TypeDef>> | Promise<Infer<ToFinalField<TypeDef>>>;

  name: string;
  kind: 'query' | 'subscription' | 'mutation';

  // keep calm ts
  typeDef: any;
  argsDef: any;
  payloadType: GraphType<any>;
  argsType: GraphType<any>;
  type: any;
  args: any;

  asObjectField(name?: string): GraphQLField<any, any>;
}

export function isPossibleArgsDef(
  args: any
): args is Readonly<ObjectDefinitionInput> {
  return args && typeof args === 'object' && Object.keys(args).length;
}
