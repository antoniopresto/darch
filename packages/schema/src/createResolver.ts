import { SchemaDefinitionInput, SchemaFieldInput } from './fields/_parseFields';
import { Infer } from './Infer';
import { createType, isSchema, parseSchemaField, Schema } from './Schema';
import { ComposeOutputType, schemaComposer } from 'graphql-compose';
import type { Resolver as GCResolver } from 'graphql-compose';
import { GraphQLFieldConfig, GraphQLResolveInfo } from 'graphql';
import { StrictMap } from '@darch/utils/lib/StrictMap';
import { fieldToGraphql, schemaToGQL } from './schemaToGQL';
import { ValidationCustomMessage } from './applyValidator';
import { schemaMetaFieldKey } from './fields/MetaFieldField';

export class Resolver<
  TypeDef extends SchemaFieldInput | Schema<SchemaDefinitionInput>,
  ArgsDef extends SchemaDefinitionInput,
  Context = unknown,
  Source = unknown
> {
  private readonly __gcResolver: GCResolver;

  constructor(
    readonly options: CreateResolverOptions<TypeDef, ArgsDef, Context, Source>
  ) {
    const { args, type, name, description, resolve, ...rest } = options;

    let parsedArgsType: Schema<any> | undefined;
    let parseArgs:
      | ((
          input: any,
          options?: { customMessage?: ValidationCustomMessage }
        ) => any)
      | undefined;

    if (isPossibleArgsDef(args)) {
      const argsSchema = createType(name, args);
      parsedArgsType = argsSchema.graphqlInputType().getFields();
      parseArgs = argsSchema.parse.bind(argsSchema);
    } else {
      parsedArgsType = undefined;
    }

    let outputType: ComposeOutputType<any>;
    let parseOutPut: (input: any) => any;

    if (isSchema(type)) {
      parseOutPut = (input) =>
        type.parse.call(type, input, {
          customMessage: (_, error) => {
            return `Invalid payload:\n ${error.message}`;
          },
        });

      outputType = type.entity();
    } else {
      const field = parseSchemaField('', type, true);

      parseOutPut = (input) =>
        field.parse.call(field, input, (_, error) => {
          return `Invalid payload:\n ${error.message}`;
        });

      if (field.type === 'schema') {
        const payloadName = field.def[schemaMetaFieldKey]?.id || name;

        const otc = schemaToGQL(payloadName, field.def);

        outputType = otc;
      } else {
        const asFinal = field.toSchemaFieldType();

        const otc = fieldToGraphql({
          field: asFinal,
          fieldName: 'Payload',
          parentName: name,
        });

        outputType = otc.type;
      }
    }

    async function finalResolve(rp: ResolveParams<any, any, any>) {
      rp.args =
        parseArgs?.(rp.args, {
          customMessage: (_, error) => {
            return `Invalid input:\n ${error.message}`;
          },
        }) || rp.args;

      const result = await resolve(rp);

      return parseOutPut(result);
    }

    this.__gcResolver = schemaComposer.createResolver({
      ...rest,
      resolve: finalResolve,
      description,
      args: parsedArgsType,
      type: outputType,
      name,
    } as any);

    Resolver.register.set(name, this);
  }

  get displayName(): string {
    return this.__gcResolver.displayName!;
  }

  set displayName(name: string) {
    this.__gcResolver.displayName = name;
  }

  getArgNames = (): string[] => {
    return this.__gcResolver.getArgNames();
  };

  resolve: this['options']['resolve'] = async (resolveParams) => {
    return this.__gcResolver.resolve(resolveParams);
  };

  getFieldConfig = <
    Extend extends GraphQLFieldConfig<Source, Context, Infer<ArgsDef>>
  >(
    extend?: Extend
  ): GraphQLFieldConfig<any, any> => {
    const fieldConfig = this.__gcResolver.getFieldConfig();

    return {
      ...fieldConfig,
      ...extend,
    };
  };

  static resetRegister = () => {
    Resolver.register.clear();
  };

  static register = new StrictMap();

  // TODO:
  // wrapResolver
  // addFilterArg
  // setArgType
  // resolver.getArgITC('filter').removeField(newFieldName);
  // hasArg
  // getArgTC
  // cloneArg
}

function isPossibleArgsDef(args: any): args is Readonly<SchemaDefinitionInput> {
  return args && typeof args === 'object' && Object.keys(args).length;
}

export function createResolver<
  TypeDef extends SchemaFieldInput | Schema<SchemaDefinitionInput>,
  ArgsDef extends SchemaDefinitionInput = { [K: string]: 'unknown?' },
  Context = unknown,
  Source = unknown
>(
  options: CreateResolverOptions<TypeDef, ArgsDef, Context, Source>
): Resolver<TypeDef, ArgsDef, Context, Source> {
  return new Resolver(options);
}

createResolver.Resolver = Resolver;

type DirectiveArgs = {
  [key: string]: any;
};

type Directive = {
  name: string;
  args?: DirectiveArgs;
};

type Extensions = {
  [key: string]: any;
  directives?: Directive[];
};

type ProjectionType = { [K: string]: any };

export interface CreateResolverOptions<
  TypeDef extends SchemaFieldInput | Schema<SchemaDefinitionInput>,
  ArgsDef extends SchemaDefinitionInput,
  Context = unknown,
  Source = unknown
> {
  type: TypeDef;
  name: string;

  resolve: (
    paramsObject: ResolveParams<
      ArgsDef extends { [K: string]: SchemaFieldInput }
        ? Infer<ArgsDef>
        : undefined,
      Context,
      Source
    >
  ) => Promise<Infer<TypeDef>>;

  args?: ArgsDef;
  kind: 'query' | 'mutation' | 'subscription';
  description?: string;
  deprecationReason?: string | null;
  projection?: ProjectionType;
  // parent?: Resolver<any, any, any, any>;
  extensions?: Extensions;
  directives?: Directive[];
}

// export type InferResolver<T> = T extends CreateResolverOptions<
//   infer Type,
//   infer Args,
//   infer Context,
//   infer Source
// >
//   ? {
//       Type: Infer<Type>;
//       Args: [Args] extends [{ [K: string]: any }] ? Infer<Args> : undefined;
//       Context: Context;
//       Source: Source;
//     }
//   : never;

export type ResolveParams<Args, Context, Source> = {
  source: Source;
  args: Args;
  context: Context;
  info: GraphQLResolveInfo;
  projection: ProjectionType;
  [opt: string]: any;
};
