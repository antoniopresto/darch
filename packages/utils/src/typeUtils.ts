// https://stackoverflow.com/questions/46176165/ways-to-get-string-literal-type-of-array-values-without-enum-overhead
export const tuple = <T extends string[]>(...args: T) => args;

export const tupleNum = <T extends number[]>(...args: T) => args;

export const tupleEnum = <T extends string[]>(
  ...values: T
): { readonly [K in T[number]]: K } => {
  return values.reduce((p, n) => {
    return {
      ...p,
      [n]: n,
    };
  }, Object.create(null));
};

export type ObjectKey<T> = T extends Record<string, any> ? T[keyof T] : never;

export type IsAny<T> = 0 extends 1 & T ? true : false;

export type IsNever<T> = [T] extends [never] ? true : false;

export type IsNullable<T> = Extract<T, null | undefined> extends never
  ? false
  : true;

export type IsOptional<T> = Extract<T, undefined> extends never ? false : true;

export type IsUnknown<T> = IsNever<T> extends false
  ? T extends unknown
    ? unknown extends T
      ? IsAny<T> extends false
        ? true
        : false
      : false
    : false
  : false;

export type OnlyKnown<T> = IsAny<T> extends true
  ? never
  : IsNever<T> extends true
  ? never
  : IsUnknown<T> extends true
  ? never
  : T;

export type MaybePromise<T> = T | Promise<T>;

export type PromiseType<P> = P extends Promise<infer T> ? T : never;

export type MaybeArray<T> = T | T[];

export type ArrayType<T> = T extends Array<infer N> ? N : T;

export type AnyRecord = Record<string, any>;

export type IfExtends<Param, Type, IfTrue, IfFalse> = Param extends Type
  ? IfTrue
  : IfFalse;

export type ObjectUnion<A, B> = {
  [K in keyof (A & B)]: (A & B)[K];
};

export type Entries<T> = {
  [K in Extract<keyof T, string>]-?: [K, T[K]];
}[Extract<keyof T, string>][];

export type AnyFunction = (...args: any[]) => any;

export type Writeable<T> = {
  -readonly [P in keyof T]: T[P];
};

export type ForceString<T> = T extends string ? T : never;
export type NotString<T> = string extends T ? never : T;

export type NullableToPartial<T> = UnionToIntersection<
  | {
      [K in keyof T as IsOptional<T[K]> extends true ? never : K]-?: T[K];
    }
  | {
      [K in keyof T as IsOptional<T[K]> extends true ? K : never]?: T[K];
    }
>;

// https://fettblog.eu/typescript-union-to-intersection/
export type UnionToIntersection<T> = (
  T extends any ? (x: T) => any : never
) extends (x: infer R) => any
  ? {
      [K in keyof R]: R[K];
    }
  : never;

export type Simplify<T> = { [KeyType in keyof T]: T[KeyType] };
