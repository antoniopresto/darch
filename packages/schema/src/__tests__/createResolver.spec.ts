import { PromiseType } from '@darch/utils/lib/typeUtils';
import { assert, IsExact } from 'conditional-type-checks';
import { GraphQLObjectType, GraphQLSchema, printSchema } from 'graphql';

import { createType, Schema } from '../Schema';
import { createResolver } from '../createResolver';

describe('createResolver', () => {
  afterEach(async () => {
    await Schema.reset();
  });

  it('Should create a Resolver', () => {
    const UserType = createType('User', {
      name: { string: {}, description: 'the user name' },
      id: 'ulid',
    });

    const resolver = createResolver({
      type: UserType,
      args: { id: 'ulid' },
      name: 'Users',
      description: 'User resolver',
      async resolve(_, { id }) {
        return {
          name: id,
          id: id,
        } as any;
      },
    });

    type Return = PromiseType<ReturnType<typeof resolver.resolve>>;
    type Args = Parameters<typeof resolver.resolve>[1];

    assert<IsExact<Return, { name: string; id: string }>>(true);
    assert<IsExact<Args, { id: string }>>(true);

    const schema = new GraphQLSchema({
      query: new GraphQLObjectType({
        name: 'Query',

        fields: () => ({
          users: resolver.fieldConfig,
        }),
      }),
    });

    expect(printSchema(schema).split('\n')).toEqual([
      'type Query {',
      '  """User resolver"""',
      '  users(id: Ulid!): User!',
      '}',
      '',
      'type User {',
      '  """the user name"""',
      '  name: String!',
      '  id: Ulid!',
      '}',
      '',
      'scalar Ulid',
    ]);
  });

  it('Should create complex types preserving names', () => {
    const user = createType('user', {
      name: 'string',
      age: 'int?',
    });

    const userAddress = createType('UserAddress', {
      street: 'string',
      number: 'int?',
    }).describe('The user address');

    const resolver = createResolver({
      type: user,
      args: {
        name: 'string',
        addresses: { schema: userAddress, list: true },
        records: { record: { keyType: 'int', type: { enum: ['banana'] } } },
      } as const,
      name: 'Users',
      resolve(rp) {
        return rp as any;
      },
    });

    type Return = ReturnType<typeof resolver.resolve>;
    type Args = Parameters<typeof resolver.resolve>[1];

    assert<IsExact<Return, Promise<{ name: string; age?: number }>>>(true);

    assert<
      IsExact<
        Args,
        {
          name: string;
          addresses: {
            number?: number | undefined;
            street: string;
          }[];
          records: { [K: number]: 'banana' };
        }
      >
    >(true);

    const schema = new GraphQLSchema({
      query: new GraphQLObjectType({
        name: 'Query',

        fields: () => ({
          users: resolver.fieldConfig,
        }),
      }),
    });

    expect(printSchema(schema).split('\n')).toEqual([
      'type Query {',
      '  users(name: String!, addresses: [UserAddressInput]!, records: UsersInput_recordsRecord!): user!',
      '}',
      '',
      'type user {',
      '  name: String!',
      '  age: Int',
      '}',
      '',
      'input UserAddressInput {',
      '  street: String!',
      '  number: Int',
      '}',
      '',
      'scalar UsersInput_recordsRecord',
    ]);
  });

  it('Should accept literals as type', () => {
    const resolver = createResolver({
      name: 'Airplanes',
      type: '[int]?',
      args: undefined,
      async resolve() {
        return undefined;
      },
    });

    const schema = new GraphQLSchema({
      query: new GraphQLObjectType({
        name: 'Query',

        fields: () => ({
          airplanes: resolver.fieldConfig,
        }),
      }),
    });

    expect(printSchema(schema).split('\n')).toEqual([
      'type Query {',
      '  airplanes: [Int]',
      '}',
    ]);
  });

  // it('Should reuse saved OTC types', async () => {
  //   const AirplaneType = createType('Airplane', {
  //     model: 'string',
  //     id: 'ulid',
  //   });
  //
  //   const UserType = createType('User', {
  //     name: 'string',
  //     collection: {
  //       schema: AirplaneType,
  //       list: true,
  //       optional: true,
  //     },
  //   });
  //
  //   const resolver = createResolver({
  //     name: 'airplanes',
  //     type: { type: AirplaneType, list: true },
  //     args: {
  //       airplanes: { type: AirplaneType, list: true },
  //       users: { schema: UserType, list: true },
  //     },
  //     async resolve() {
  //       return [];
  //     },
  //   });
  //
  //   const schema = new GraphQLSchema({
  //     query: new GraphQLObjectType({
  //       name: 'Query',
  //
  //       fields: () => ({
  //         airplanes: resolver.getFieldConfig(),
  //         airplanes2: resolver.getFieldConfig(),
  //         airplaneFromSchema: AirplaneType.createResolver({
  //           name: 'airplane',
  //           args: { id: 'int' },
  //           resolve() {
  //             return {} as any;
  //           },
  //         }).getFieldConfig(),
  //
  //         user: UserType.createResolver({
  //           name: 'user',
  //           args: { id: 'int' },
  //           resolve() {
  //             return {} as any;
  //           },
  //         }).getFieldConfig(),
  //
  //         users: Schema.resolverFieldConfig({
  //           name: 'users',
  //           type: { schema: UserType, list: true },
  //           args: {},
  //           resolve() {
  //             return {} as any;
  //           },
  //         }),
  //
  //         airplaneByAP: UserType.createResolver({
  //           name: 'airplaneByAP',
  //           args: { ap: AirplaneType },
  //           resolve() {
  //             return {} as any;
  //           },
  //         }).getFieldConfig(),
  //       }),
  //     }),
  //   });
  //
  //   expect(printSchema(schema)).toEqual('');
  // });

  // it('Should reuse arg types', async () => {
  //   const name = 'Salad';
  //
  //   const Fruit = createType('Fruit', {
  //     name: 'string',
  //     color: 'string',
  //   });
  //
  //   const a = Resolver.argsToGQL({
  //     name,
  //     args: {
  //       name: 'string',
  //       fruits: { list: true, type: Fruit },
  //     },
  //   });
  //
  //   expect(a.result).toEqual({});
  // });
});
