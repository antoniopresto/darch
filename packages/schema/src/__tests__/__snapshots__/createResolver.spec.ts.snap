// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createResolver Should accept literals as type 1`] = `
"type Query {
  airplanes: [Int]
}
"
`;

exports[`createResolver Should create complex types preserving names 1`] = `
"scalar FIXMEUnionFIXME

type Query {
  users(id: Ulid!, name: String!, addresses: UserAddressInput, records: Record!): user
}

scalar Record

scalar Ulid

type user {
  name: String!
  age: Int
}

input UserAddressInput {
  street: String!

  \\"\\"\\"string or number\\"\\"\\"
  number: FIXMEUnionFIXME!
}
"
`;
